### ISSUE TITLE
storage: a row sometimes is not returned although it is present in DB

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#37516]
While running [Roachprod based release qualification](https://wiki.crdb.io/wiki/spaces/CRDB/pages/77955117/Roachprod+based+release+qualification) for 2.1.7 release, the TPCC workload on one of the clusters encountered `Error: error in newOrder: missing stock row`. I added a simple modification to `workload` code to see where the error is coming from as [here](https://gist.github.com/yuzefovich/f44b203df124933af0ddbbc443f98e8e).

(Note: going forward in this issue all the examples will be from `yahor-release-v216-2` roachprod cluster that currently will be live for another 4 days or so.)

By slightly modifying the query that errors out, we find that a row with primary key `(s_w_id, s_i_id) = (210, 78206)` is missing from the output.

The behavior is interesting, however:
- if this key is a part of the first span to be scanned, the corresponding row is returned correctly:
```
root@localhost:26257/tpcc> SELECT s_i_id, s_quantity FROM stock WHERE (s_i_id, s_w_id) IN ((78205, 210), (78206, 210));
  s_i_id | s_quantity  
+--------+------------+
   78205 |         13  
   78206 |         34  
(2 rows)
```
 - if the key is not the part of the first span, then the row is omitted:
```
root@localhost:26257/tpcc> SELECT s_i_id, s_quantity FROM stock WHERE (s_i_id, s_w_id) IN ((78204, 210), (78206, 210));
  s_i_id | s_quantity  
+--------+------------+
   78204 |         61  
(1 row)
```
- if we're asking only for columns in the primary index, then a different index is used for the scan, and the row is returned:
```
root@localhost:26257/tpcc> SELECT s_w_id, s_i_id FROM stock WHERE (s_i_id, s_w_id) IN ((78204, 210), (78206, 210));
  s_w_id | s_i_id  
+--------+--------+
     210 |  78204  
     210 |  78206  
(2 rows)

Time: 1.337604ms

root@localhost:26257/tpcc> EXPLAIN SELECT s_w_id, s_i_id FROM stock WHERE (s_i_id, s_w_id) IN ((78204, 210), (78206, 210));
    tree    | field |                 description                  
+-----------+-------+---------------------------------------------+
  render    |       |                                              
   └── scan |       |                                              
            | table | stock@stock_s_i_id_idx                       
            | spans | /78204/210-/78204/211 /78206/210-/78206/211  
(4 rows)

Time: 757.316µs

root@localhost:26257/tpcc> EXPLAIN SELECT s_w_id, s_i_id, s_quantity FROM stock WHERE (s_i_id, s_w_id) IN ((78204, 210), (78206, 210));
    tree    | field |                   description                    
+-----------+-------+-------------------------------------------------+
  render    |       |                                                  
   └── scan |       |                                                  
            | table | stock@primary                                    
            | spans | /210/78204-/210/78204/# /210/78206-/210/78206/#  
(4 rows)
```

After some debugging, it seems that the problem is somewhere below the SQL layer, and Jordan printed out some useful traces [here](https://gist.github.com/jordanlewis/a5226bc64bc19497935be69d96eeea72).

I think I've hit this error 4 times out of 10 attempts on 2.1 releases (I haven't tried reproducing on 19.1). To reproduce simply follow [Running a workload on 2.1 clusters](https://wiki.crdb.io/wiki/spaces/CRDB/pages/77955117/Roachprod+based+release+qualification) from the bottom of the page. At the moment, `yahor-release-v216-2` and `yahor-release-v217-2` have this problem (the former with `(s_w_id, s_i_id) = (210, 78206)`, and the latter with `(s_w_id, s_i_id) = (705, 41244)`). The error occurred (at least so far) only on `stock` table. Note that restarting the nodes seemed to have resolved the issue.

**Comment by ajkr** [comment]
https://s3.us-east-2.amazonaws.com/andrewk-artifacts/debug-archive.tar.gz

- The query that finds one row where two are expected involves: (1) `Seek` targeting a key beginning with `0xcd89f7016af801453e`, then (2) `Seek` targeting a key beginning with `0xcd89f7016af8015ef3`. If we run the queries separately each returns one row. If we run in a single SQL select, we see only the row for (1).
- "002107.sst" spans range `['CD89F7016AF79CF38800159F35086A2A09D009' seq:7900408, type:1 .. 'CD89F7016AF8015EF38800159F35086A2A09D009' seq:72057594037927935, type:15]`
- "002136.sst" spans range `['CD89F7016AF8015EF38800159F35086A2A09D009' seq:7900408, type:1 .. 'CD89F7016BF79A588800159F35086A2A09D009' seq:72057594037927935, type:15]`
- There is a range tombstone overlapping the two files with the range `[0xCD89F70168F801522A00, CD89F70171F8017FEC00)`
- The key found for (1) is `0xCD89F7016AF801453E8800159F35086A2A09D009` in "002107.sst"
- The key that should be found for (2) is `0xCD89F7016AF8015EF38800159F35086A2A09D009` in "002136.sst" (it is the first key)

**Comment by bdarnell** [comment]
Each sst has its own range tombstone. 

```
002107_dump.txt-    HEX    CD89F70168F801522A00: CD89F70171F8017FEC00
002136_dump.txt-    HEX    CD89F7016AF8015EF38800159F35086A2A09D009: CD89F70171F799BD8800159F35086A2A09D009
```

Putting things in order, we see that the range tombstone in the first sst (2107) covers both ssts (and then some), while the second sst doesn't know about that tombstone but does have a second smaller range tombstone. 

```
CD89F70168F801522A00 sst 1 range tombstone start
CD89F7016AF79CF38800159F35086A2A09D009 sst 1 manifest start
CD89F7016AF79D528800159F35086A2A09D009 sst 1 first key
CD89F7016AF801453E8800159F35086A2A09D009 first queried key
CD89F7016AF8015EF28800159F35086A2A09D009 sst 1 last key
CD89F7016AF8015EF38800159F35086A2A09D009 second queried key
CD89F7016AF8015EF38800159F35086A2A09D009 sst 1 manifest end
CD89F7016AF8015EF38800159F35086A2A09D009 sst 2 manifest start
CD89F7016AF8015EF38800159F35086A2A09D009 sst 2 range tombstone start
CD89F7016AF8015F538800159F35086A2A09D009 sst 2 first key
CD89F7016BF79A578800159F35086A2A09D009 sst 2 last key
CD89F7016BF79A588800159F35086A2A09D009 sst 2 manifest end
CD89F70171F799BD8800159F35086A2A09D009 sst 2 range tombstone end
CD89F70171F8017FEC00 sst 1 range tombstone end
```

Decoding these keys, they look like:

```
$ cut -f1 -d' ' /tmp/foo.txt|xargs  cockroach debug decode-key | paste - <(cut -f2- -d' ' /tmp/foo.txt)
```

```
/Table/69/1/360/86570	sst 1 range tombstone start
/Table/69/1/362/40179/0/1558022306.350631376,0	sst 1 manifest start
/Table/69/1/362/40274/0/1558022306.350631376,0	sst 1 first key
/Table/69/1/362/83262/0/1558022306.350631376,0	first queried key
/Table/69/1/362/89842/0/1558022306.350631376,0	sst 1 last key
/Table/69/1/362/89843/0/1558022306.350631376,0	second queried key
/Table/69/1/362/89843/0/1558022306.350631376,0	sst 1 manifest end
/Table/69/1/362/89843/0/1558022306.350631376,0	sst 2 manifest start
/Table/69/1/362/89843/0/1558022306.350631376,0	sst 2 range tombstone start
/Table/69/1/362/89939/0/1558022306.350631376,0	sst 2 first key
/Table/69/1/363/39511/0/1558022306.350631376,0	sst 2 last key
/Table/69/1/363/39512/0/1558022306.350631376,0	sst 2 manifest end
/Table/69/1/369/39357/0/1558022306.350631376,0	sst 2 range tombstone end
/Table/69/1/369/98284	sst 1 range tombstone end
```

The fact that the first sst's range tombstone's keys are so much shorter than the others (not including timestamps) indicates that they are somehow derived from a manual split instead of natural size-based splitting. 

I'm not sure what, if anything, all this can tell us. It makes me vaguely suspicious of the work we did around range tombstones (and I'm not sure if all of this made it upstream, so it could be cockroach-specific patches). For example, https://github.com/cockroachdb/rocksdb/commit/baf0248e669377b4b195460795f40b99f58222d0 has to do with the truncation of range tombstones at sst boundaries and touches comments like "the discrepancy between FileMetaData::largest being inclusive while RangeTombstone::end_key_ is exclusive".

**Comment by ajkr** [comment]
One thing both Yahor and I noticed is the problem tends to happen either shortly after import or not at all. For me, every time it happened was within 20 minutes of starting TPCC after import, and every time it didn't happen it ran for 2-4 hours with no problem (until I killed it).

Maybe the new workload generates files with large range tombstones that only span multiple files upon compaction to a lower level? Just guessing here. I didn't notice anything particularly suspicious in the info we have.

**Comment by ajkr** [comment]
Also it might be possible the two range tombstones you see are the same one. When a range tombstone spans a file, the range start key in the second file gets truncated to the smallest point key in that file, which appears to have happened. The differing end keys is harder to explain -- I cannot think of what truncation logic caused them to differ like that, but perhaps there is something.

**Comment by yuzefovich** [comment]
I did see an error appear after significant time running a workload - once after about 24 hrs and another time after about 16 hrs.

**Comment by bdarnell** [comment]
This can be closed now that #37694 is merged, right?

**Comment by ajkr** [comment]
Yes.
